# Permutations

In the previous notebook we learned the **product rule** and **sum rule** of counting, and saw how naive counting can lead to **double counting**. In this notebook we tackle a specific and very common counting question:

> **In how many ways can we arrange (order) objects?**

We will:
1. Derive the permutation formula from the product rule
2. Handle the case where **all objects are distinct**
3. Handle the case where **some objects are repeated** (not all distinct)
4. Work through examples and practice problems

In [44]:
import math
from itertools import permutations

print("Libraries imported successfully!")

Libraries imported successfully!


---
## 1. Motivating the Problem

Suppose we have 3 books — **A**, **B**, **C** — and we want to arrange them on a shelf.

Using the **product rule**:
- Position 1: **3** choices (any of the 3 books)
- Position 2: **2** choices (whichever 2 remain)
- Position 3: **1** choice (the last book)

$$\text{Total arrangements} = 3 \times 2 \times 1 = 6$$

Let's verify by listing them all:

In [45]:
books = ["A", "B", "C"]

all_arrangements = list(permutations(books))

print(f"Books: {books}")
print(f"Total arrangements: {len(all_arrangements)}")
print()
for i, arr in enumerate(all_arrangements, 1):
    print(f"  {i}. {' → '.join(arr)}")

Books: ['A', 'B', 'C']
Total arrangements: 6

  1. A → B → C
  2. A → C → B
  3. B → A → C
  4. B → C → A
  5. C → A → B
  6. C → B → A


---
## 2. Deriving the Permutation Formula (All Distinct)

### Arranging all $n$ distinct objects

Generalizing the book example, if we have $n$ **distinct** objects and want to arrange **all** of them:

| Position | Choices available |
|----------|------------------|
| 1st | $n$ |
| 2nd | $n - 1$ |
| 3rd | $n - 2$ |
| ... | ... |
| $n$-th | $1$ |

By the **product rule** (each position is a sequential step):

$$n \times (n-1) \times (n-2) \times \cdots \times 2 \times 1 = n!$$

This product is called **$n$ factorial**, written $n!$

### Definition

$$n! = \prod_{i=1}^{n} i = 1 \times 2 \times 3 \times \cdots \times n$$

By convention: $0! = 1$

In [46]:
# Computing factorials
for n in range(8):
    # Build the multiplication string: 1 × 2 × ... × n
    if n == 0:
        expr = "1 (by convention)"
    else:
        parts = []
        for i in range(1, n + 1):
            parts.append(str(i))
        expr = " × ".join(parts)
    print(f"{n}! = {expr} = {math.factorial(n)}")

0! = 1 (by convention) = 1
1! = 1 = 1
2! = 1 × 2 = 2
3! = 1 × 2 × 3 = 6
4! = 1 × 2 × 3 × 4 = 24
5! = 1 × 2 × 3 × 4 × 5 = 120
6! = 1 × 2 × 3 × 4 × 5 × 6 = 720
7! = 1 × 2 × 3 × 4 × 5 × 6 × 7 = 5040


Notice how quickly factorials grow — this is why brute-force enumeration becomes impractical even for moderate $n$.

In [47]:
# Factorial growth
for n in [10, 15, 20, 25, 52]:
    print(f"{n:>2}! = {math.factorial(n):,}")

print()
print("52! is the number of ways to shuffle a standard deck of cards.")
print(f"It has {len(str(math.factorial(52)))} digits!")

10! = 3,628,800
15! = 1,307,674,368,000
20! = 2,432,902,008,176,640,000
25! = 15,511,210,043,330,985,984,000,000
52! = 80,658,175,170,943,878,571,660,636,856,403,766,975,289,505,440,883,277,824,000,000,000,000

52! is the number of ways to shuffle a standard deck of cards.
It has 68 digits!


---
## 3. Choosing and Arranging $r$ out of $n$ — $P(n, r)$

Often we don't arrange **all** $n$ objects — we pick **$r$** of them and arrange those.

### Derivation using the Product Rule

Pick $r$ objects from $n$ distinct objects, **order matters**:

| Position | Choices |
|----------|--------|
| 1st | $n$ |
| 2nd | $n - 1$ |
| 3rd | $n - 2$ |
| ... | ... |
| $r$-th | $n - r + 1$ |

By the product rule:

$$P(n, r) = n \times (n-1) \times (n-2) \times \cdots \times (n - r + 1)$$

This is a product of $r$ consecutive descending integers starting from $n$.

### Compact formula using factorials

$$P(n, r) = \frac{n!}{(n - r)!}$$

**Why?** Because $n! = n \times (n-1) \times \cdots \times (n-r+1) \times (n-r)!$, so dividing by $(n-r)!$ cancels the unwanted tail:

$$\frac{n!}{(n-r)!} = \frac{\cancel{1 \times 2 \times \cdots \times (n-r)} \times (n-r+1) \times \cdots \times n}{\cancel{1 \times 2 \times \cdots \times (n-r)}}$$

### Special cases
- $P(n, n) = \frac{n!}{0!} = n!$ — arranging all $n$ objects
- $P(n, 1) = n$ — just picking one object
- $P(n, 0) = 1$ — doing nothing (one way to do nothing)

In [48]:
def P(n, r):
    """Compute P(n, r) = n! / (n - r)!"""
    return math.factorial(n) // math.factorial(n - r)

# Verify: P(n, r) = product of r terms
def P_product(n, r):
    """Compute P(n, r) using the product rule directly."""
    result = 1
    for i in range(r):
        result *= (n - i)
    return result

# Show both methods agree
print("n    r    P(n,r) [factorial]    P(n,r) [product]    Match?")
print("-" * 62)
for n, r in [(5, 3), (6, 2), (8, 4), (10, 5), (4, 4), (7, 1), (5, 0)]:
    p_fact = P(n, r)
    p_prod = P_product(n, r)
    match = "✓" if p_fact == p_prod else "✗"
    print(f"{n:<5}{r:<5}{p_fact:<22}{p_prod:<22}{match}")

n    r    P(n,r) [factorial]    P(n,r) [product]    Match?
--------------------------------------------------------------
5    3    60                    60                    ✓
6    2    30                    30                    ✓
8    4    1680                  1680                  ✓
10   5    30240                 30240                 ✓
4    4    24                    24                    ✓
7    1    7                     7                     ✓
5    0    1                     1                     ✓


### Example 1: Race Podium

In a race with **10 runners**, how many ways can the **gold, silver, and bronze** medals be awarded?

We are choosing 3 from 10, and **order matters** (gold ≠ silver ≠ bronze).

$$P(10, 3) = 10 \times 9 \times 8 = 720$$

In [49]:
n, r = 10, 3
result = P(n, r)

print(f"Runners: {n}")
print(f"Medal positions: {r} (gold, silver, bronze)")
print(f"P({n}, {r}) = {n} × {n-1} × {n-2} = {result}")

Runners: 10
Medal positions: 3 (gold, silver, bronze)
P(10, 3) = 10 × 9 × 8 = 720


### Example 2: Seating Arrangement

There are **7 people** and **4 chairs** in a row. How many seating arrangements are possible?

$$P(7, 4) = 7 \times 6 \times 5 \times 4 = 840$$

In [50]:
n, r = 7, 4
result = P(n, r)

parts = []
for i in range(r):
    parts.append(str(n - i))
terms = " × ".join(parts)

print(f"People: {n}, Chairs: {r}")
print(f"P({n}, {r}) = {terms} = {result}")
print()
print(f"Using factorial formula: {n}! / ({n}-{r})! = {math.factorial(n)} / {math.factorial(n-r)} = {result}")

People: 7, Chairs: 4
P(7, 4) = 7 × 6 × 5 × 4 = 840

Using factorial formula: 7! / (7-4)! = 5040 / 6 = 840


### Example 3: Word Arrangements

How many 3-letter "words" (not necessarily meaningful) can be formed from the letters of **PYTHON** (6 distinct letters), if **no letter is repeated**?

$$P(6, 3) = 6 \times 5 \times 4 = 120$$

In [51]:
word = "PYTHON"
r = 3
n = len(word)

result = P(n, r)
print(f"Letters: {list(word)} ({n} distinct)")
print(f"Choosing {r} letters, order matters, no repetition")
print(f"P({n}, {r}) = {result}")
print()

# Show a few of them
sample = list(permutations(word, r))
print(f"First 10 of {len(sample)} arrangements:")
for i, arr in enumerate(sample[:10], 1):
    print(f"  {i:>2}. {''.join(arr)}")
print(f"  ...")

Letters: ['P', 'Y', 'T', 'H', 'O', 'N'] (6 distinct)
Choosing 3 letters, order matters, no repetition
P(6, 3) = 120

First 10 of 120 arrangements:
   1. PYT
   2. PYH
   3. PYO
   4. PYN
   5. PTY
   6. PTH
   7. PTO
   8. PTN
   9. PHY
  10. PHT
  ...


### Example 4: Smudge Attack on a 6-Digit Passcode

You notice **6 smudge marks** on someone's phone screen, each on a **different digit**. The passcode is **6 digits** long.

Since there are 6 smudges and 6 digits in the passcode, **every digit is used exactly once** — all 6 digits are distinct.

An attacker only needs to figure out the **order**. This is a permutation of 6 distinct items:

$$P(6, 6) = 6! = 720$$

That's only 720 attempts — very feasible to brute-force!

**What if only 3 smudges were found?** Then only 3 distinct digits are used across 6 positions, meaning some digits repeat. The exact count depends on how many times each digit repeats — this is where **permutations with repetition** (Section 4) come in.

In [52]:
# 6-digit passcode with 6 distinct smudges
smudges = 6
passcode_length = 6

# All digits are distinct, so we just need the order
possible_passcodes = P(smudges, passcode_length)

print(f"Smudge marks found: {smudges} (all different digits)")
print(f"Passcode length: {passcode_length}")
print(f"Possible passcodes: P({smudges}, {passcode_length}) = {smudges}! = {possible_passcodes}")
print()

# Compare with no information (any digit 0-9 in each position)
no_info = 10 ** passcode_length
print(f"Without smudge info: 10⁶ = {no_info:,}")
print(f"With smudge info:    6!  = {possible_passcodes}")
print(f"Search space reduced by: {no_info // possible_passcodes}x")

Smudge marks found: 6 (all different digits)
Passcode length: 6
Possible passcodes: P(6, 6) = 6! = 720

Without smudge info: 10⁶ = 1,000,000
With smudge info:    6!  = 720
Search space reduced by: 1388x


### Example 5: Smudge Attack — Only 5 Smudges on a 6-Digit Passcode

Now suppose you find only **5 smudge marks**. The passcode is still **6 digits** long.

Since 5 distinct digits fill 6 positions, **exactly one digit is used twice** and the other four are used once. The frequency pattern is always $(2, 1, 1, 1, 1)$.

This is a direct application of the **permutations with repetition** formula:

$$\frac{n!}{n_1! \times n_2! \times n_3! \times n_4! \times n_5!} = \frac{6!}{2! \times 1! \times 1! \times 1! \times 1!} = \frac{720}{2} = 360$$

But **which** of the 5 digits is the repeated one? Each choice gives the same frequency pattern, so by the **sum rule**:

$$\text{Total} = 5 \times \frac{6!}{2! \times 1! \times 1! \times 1! \times 1!} = 5 \times 360 = 1800$$

In [ ]:
# 6-digit passcode with 5 distinct smudges
smudges = 5
passcode_length = 6

# Frequency pattern: (2, 1, 1, 1, 1) — one digit repeated, rest appear once
# Permutations with repetition: n! / (n1! × n2! × ... × nk!)
# Only the repeated digit contributes to the denominator (1! = 1 for the rest)
arrangements_per_pattern = math.factorial(passcode_length) // math.factorial(2)

# Which digit is repeated? 5 choices → sum rule
total = smudges * arrangements_per_pattern

print(f"Smudge marks found: {smudges}")
print(f"Passcode length: {passcode_length}")
print(f"Frequency pattern: (2, 1, 1, 1, 1)")
print()
print(f"Permutations with repetition: {passcode_length}! / (2! × 1! × 1! × 1! × 1!) = {math.factorial(passcode_length)} / {math.factorial(2)} = {arrangements_per_pattern}")
print(f"Which digit repeats? {smudges} choices (sum rule)")
print(f"Total: {smudges} × {arrangements_per_pattern} = {total}")
print()

# Compare
no_info = 10 ** passcode_length
six_smudge = math.factorial(6)
print(f"No smudge info: {no_info:,}")
print(f"6 smudges:      {six_smudge}")
print(f"5 smudges:      {total}")
print(f"Search space reduced by: {no_info // total}x (vs no info)")

---
## 4. Permutations with Repeated Elements

### The Problem

What if not all objects are distinct? Consider the word **BANANA**:
- 6 letters total
- But: B appears **1** time, A appears **3** times, N appears **2** times

If all 6 letters were distinct, there would be $6! = 720$ arrangements. But since some letters are **identical**, many of those 720 arrangements look the same.

### Why $n!$ overcounts

Label the A's as $A_1, A_2, A_3$ and N's as $N_1, N_2$. Then:
- $B A_1 N_1 A_2 N_2 A_3$
- $B A_2 N_1 A_1 N_2 A_3$
- $B A_1 N_2 A_2 N_1 A_3$
- ... (and more)

All of these look like **BANANA** when we drop the subscripts. We **overcounted** because swapping identical items among themselves produces the same arrangement.

### Correcting the overcount

- The 3 identical A's can be permuted among themselves in $3!$ ways, all producing the same word.
- The 2 identical N's can be permuted in $2!$ ways.
- The 1 B can be permuted in $1!$ way.

Each **distinct** arrangement was counted $3! \times 2! \times 1! = 12$ times.

### Formula: Permutations with Repetition

If we have $n$ objects where object type $i$ appears $n_i$ times (and $n_1 + n_2 + \cdots + n_k = n$):

$$\text{Distinct permutations} = \frac{n!}{n_1! \times n_2! \times \cdots \times n_k!}$$

This is sometimes called a **multinomial coefficient**.

### Derivation

1. Start: if all $n$ were distinct → $n!$ arrangements (product rule)
2. But swapping identical items of type $i$ among themselves ($n_i!$ ways) gives the same arrangement
3. By the product rule, the total overcount factor is $n_1! \times n_2! \times \cdots \times n_k!$
4. Divide to correct: $\frac{n!}{n_1! \times n_2! \times \cdots \times n_k!}$

### Example 6: Arrangements of BANANA

$$\frac{6!}{3! \times 2! \times 1!} = \frac{720}{6 \times 2 \times 1} = \frac{720}{12} = 60$$

In [53]:
from collections import Counter

def permutations_with_repetition(items):
    """Count distinct permutations when items may repeat."""
    n = len(items)
    counts = Counter(items)
    numerator = math.factorial(n)
    denominator = 1
    for c in counts.values():
        denominator *= math.factorial(c)
    return numerator // denominator, n, counts

word = "BANANA"
result, n, counts = permutations_with_repetition(word)

print(f"Word: {word}")
print(f"Length: {n}")
print(f"Letter frequencies: {dict(counts)}")
print()

denom_part_list = []
for c in counts.values():
    denom_part_list.append(f"{c}!")
denom_parts = " × ".join(denom_part_list)

denom_value_list = []
for c in counts.values():
    denom_value_list.append(str(math.factorial(c)))
denom_values = " × ".join(denom_value_list)

denom = 1
for c in counts.values():
    denom *= math.factorial(c)

print(f"Formula: {n}! / ({denom_parts})")
print(f"       = {math.factorial(n)} / ({denom_values})")
print(f"       = {math.factorial(n)} / {denom}")
print(f"       = {result}")

Word: BANANA
Length: 6
Letter frequencies: {'B': 1, 'A': 3, 'N': 2}

Formula: 6! / (1! × 3! × 2!)
       = 720 / (1 × 6 × 2)
       = 720 / 12
       = 60


In [54]:
# Verify by brute force: generate all permutations and count unique ones
all_perms = set(permutations(word))
print(f"Brute force unique permutations of '{word}': {len(all_perms)}")
print(f"Formula result: {result}")
print(f"Match: {'✓' if len(all_perms) == result else '✗'}")
print()
print(f"Total permutations (treating all as distinct): {math.factorial(n)}")
print(f"Overcount factor: {math.factorial(n) // result}")

Brute force unique permutations of 'BANANA': 60
Formula result: 60
Match: ✓

Total permutations (treating all as distinct): 720
Overcount factor: 12


### Example 7: Arrangements of MISSISSIPPI

- M: 1, I: 4, S: 4, P: 2 → total 11 letters

$$\frac{11!}{1! \times 4! \times 4! \times 2!} = \frac{39916800}{1 \times 24 \times 24 \times 2} = \frac{39916800}{1152} = 34650$$

In [55]:
word = "MISSISSIPPI"
result, n, counts = permutations_with_repetition(word)

print(f"Word: {word}")
print(f"Length: {n}")
print(f"Letter frequencies: {dict(counts)}")
print()

denom_part_list = []
for c in counts.values():
    denom_part_list.append(f"{c}!")
denom_parts = " × ".join(denom_part_list)

denom = 1
for c in counts.values():
    denom *= math.factorial(c)

print(f"Formula: {n}! / ({denom_parts})")
print(f"       = {math.factorial(n):,} / {denom:,}")
print(f"       = {result:,}")

Word: MISSISSIPPI
Length: 11
Letter frequencies: {'M': 1, 'I': 4, 'S': 4, 'P': 2}

Formula: 11! / (1! × 4! × 4! × 2!)
       = 39,916,800 / 1,152
       = 34,650


### Example 8: Comparing Distinct vs Repeated

Let's see how repetition affects the count across a few words:

In [56]:
words = ["ABCDEF", "AABBCC", "AAAAAA", "ABCABC", "PEPPER"]

print(f"{'Word':<12} {'Length':>6} {'If all distinct':>16} {'With repetition':>16} {'Overcount':>10}")
print("-" * 64)

for word in words:
    result, n, counts = permutations_with_repetition(word)
    all_distinct = math.factorial(n)
    overcount = all_distinct // result
    print(f"{word:<12} {n:>6} {all_distinct:>16,} {result:>16,} {overcount:>10,}")

Word         Length  If all distinct  With repetition  Overcount
----------------------------------------------------------------
ABCDEF            6              720              720          1
AABBCC            6              720               90          8
AAAAAA            6              720                1        720
ABCABC            6              720               90          8
PEPPER            6              720               60         12


Notice:
- **ABCDEF** (all distinct): overcount = 1, no correction needed
- **AAAAAA** (all same): only 1 distinct arrangement regardless of length
- More repetition → larger overcount → fewer distinct arrangements

---
## 5. Summary of Formulas

| Scenario | Formula | Example |
|----------|---------|--------|
| Arrange all $n$ distinct objects | $n!$ | 5 books on shelf: $5! = 120$ |
| Arrange $r$ of $n$ distinct objects | $P(n,r) = \frac{n!}{(n-r)!}$ | 3 medals from 10 runners: $P(10,3) = 720$ |
| Arrange all $n$ objects, some repeated | $\frac{n!}{n_1! \cdot n_2! \cdots n_k!}$ | BANANA: $\frac{6!}{3!2!1!} = 60$ |

---
## 6. Practice Problems

Try solving these on your own before running the solution cells!

### Problem 1: Committee Positions

A club has **12 members**. They need to elect a **President**, **Vice President**, and **Secretary**.

How many ways can these positions be filled? (No one can hold two positions.)

In [57]:
# SOLUTION
# Order matters (President ≠ Vice President), so this is P(12, 3)
n, r = 12, 3
result = P(n, r)

print(f"P({n}, {r}) = {n} × {n-1} × {n-2} = {result}")

P(12, 3) = 12 × 11 × 10 = 1320


### Problem 2: Distinct Arrangements of MATHEMATICS

How many distinct arrangements of the letters in **MATHEMATICS** are there?

Hint: First count the frequency of each letter.

In [58]:
# SOLUTION
word = "MATHEMATICS"
result, n, counts = permutations_with_repetition(word)

print(f"Word: {word}")
print(f"Length: {n}")
print(f"Frequencies: {dict(counts)}")
print()

# M=2, A=2, T=2, H=1, E=1, I=1, C=1, S=1
denom_part_list = []
for c in counts.values():
    denom_part_list.append(f"{c}!")
denom_parts = " × ".join(denom_part_list)

print(f"Formula: {n}! / ({denom_parts}) = {result:,}")

Word: MATHEMATICS
Length: 11
Frequencies: {'M': 2, 'A': 2, 'T': 2, 'H': 1, 'E': 1, 'I': 1, 'C': 1, 'S': 1}

Formula: 11! / (2! × 2! × 2! × 1! × 1! × 1! × 1! × 1!) = 4,989,600


### Problem 3: Digit Arrangements

How many distinct **7-digit numbers** can be formed using the digits **1, 1, 2, 2, 2, 3, 3**?

In [59]:
# SOLUTION
digits = [1, 1, 2, 2, 2, 3, 3]
result, n, counts = permutations_with_repetition(digits)

print(f"Digits: {digits}")
print(f"Frequencies: {dict(counts)}")
print(f"Distinct arrangements: {n}! / (2! × 3! × 2!) = {result}")

Digits: [1, 1, 2, 2, 2, 3, 3]
Frequencies: {1: 2, 2: 3, 3: 2}
Distinct arrangements: 7! / (2! × 3! × 2!) = 210


### Problem 4: Ranking Students

In a class of **20 students**, how many ways can the **top 5** positions be assigned in an exam? (No ties.)

In [60]:
# SOLUTION
# Order matters (1st ≠ 2nd), choosing 5 from 20
n, r = 20, 5
result = P(n, r)

parts = []
for i in range(r):
    parts.append(str(n - i))
terms = " × ".join(parts)

print(f"P({n}, {r}) = {terms} = {result:,}")

P(20, 5) = 20 × 19 × 18 × 17 × 16 = 1,860,480


### Problem 5: Arranging Books on a Shelf

You have **3 identical math books**, **2 identical science books**, and **4 identical history books** (9 books total).

In how many **distinct** ways can they be arranged on a shelf?

In [61]:
# SOLUTION
# 9 books: M=3, S=2, H=4
n = 9
math_books = 3
science_books = 2
history_books = 4

result = math.factorial(n) // (math.factorial(math_books) * math.factorial(science_books) * math.factorial(history_books))

print(f"Total books: {n} (Math: {math_books}, Science: {science_books}, History: {history_books})")
print(f"Arrangements: {n}! / ({math_books}! × {science_books}! × {history_books}!)")
print(f"            = {math.factorial(n)} / ({math.factorial(math_books)} × {math.factorial(science_books)} × {math.factorial(history_books)})")
print(f"            = {result:,}")

Total books: 9 (Math: 3, Science: 2, History: 4)
Arrangements: 9! / (3! × 2! × 4!)
            = 362880 / (6 × 2 × 24)
            = 1,260


### Problem 6: Strings from Letters

How many **4-letter strings** can be formed from the letters **A, B, C, D, E** if:

**Part A**: No repetition allowed?

**Part B**: Repetition is allowed?

In [62]:
# SOLUTION
n = 5  # letters A-E
r = 4  # string length

# Part A: No repetition → P(n, r)
part_a = P(n, r)
print(f"Part A (no repetition):  P({n}, {r}) = {n} × {n-1} × {n-2} × {n-3} = {part_a}")

# Part B: Repetition allowed → n^r (product rule, n choices at each of r positions)
part_b = n ** r
print(f"Part B (with repetition): {n}⁴ = {part_b}")
print()
print(f"Repetition adds {part_b - part_a} more possibilities.")

Part A (no repetition):  P(5, 4) = 5 × 4 × 3 × 2 = 120
Part B (with repetition): 5⁴ = 625

Repetition adds 505 more possibilities.


---
## 7. What's Next?

Permutations count **ordered** arrangements. But many problems ask:

> *How many ways can we **choose** $r$ items from $n$, where order does **not** matter?*

For example, choosing 3 people from 10 for a committee (no positions) — the group {Alice, Bob, Carol} is the same as {Carol, Alice, Bob}.

This leads to **combinations**, which we derive by correcting permutations for the overcount due to ordering:

$$C(n, r) = \frac{P(n, r)}{r!} = \frac{n!}{r!(n-r)!}$$

→ *Next notebook: Combinations*